### ISSUE TITLE
Subscription conflicts due to statics in generic class

### ISSUE THREAD

**Issue by scottt732** [PeteGoo/Pushqa#5]
Your subscriptions ConcurrentDictionary is currently at the QueryablePushService<T> class level.  I think I stumbled on an issue with this.  Take an app with 2 connections mapped.  If a single client tries to subscribe to the two push service items (with the same connectionId), I think one subscription replaces the other.
app.MapConnection<QueryablePushService<EventContext>>("events", connConfig);
app.MapConnection<QueryablePushService<LogContext>>("log", connConfig);

[According to ReSharper...] Static objects in generic classes are shared regardless of the type of T.  So QueryablePushService<EventContext> and QueryablePushService<LogContext> are sharing the same static instance of the dictionary, though it might not seem intuitive.  I don't think that's what you intended.  Subscribing to one while already subscribed to the other would cause the TryAddOrUpdate to run the update routine, replacing the other subscription.

I think you need to explicitly track {connectionId,subscription} on a per-type of T basis.  I think this would need to go in a non-generic base class (e.g., QueryablePushService).  Something along the lines of Dictionary<Type, ConcurrentDictionary<string, IDisposable>>.

Also, I think you need to make sure that you're disposing the IDisposables/subscriptions in case they are infinite sequences.  So capture the out param on the TryRemove and call Dispose on it, and also modify your AddOrUpdate calls so that the third parameter disposes the one it's replacing.  (x, y) => subscription.

**Comment by PeteGoo** [comment]
Thanks for the feedback. 

In regards to static state on generic types see this gist I just created https://gist.github.com/PeteGoo/5043985  
also this http://stackoverflow.com/questions/9647641/resharper-warns-static-field-in-generic-type

Totally agree on the cleanup of the subscriptions. I'll have a look at a fix soon.

**Comment by IanYates** [comment]
The gist looks like it confirms Jon Skeet's answer on StackOverflow, which disagrees with 

> [According to ReSharper...] Static objects in generic classes are shared regardless of the type of T.

There's a bit on the gist which threw me for a moment when I adapted it for Linqpad - You had

>    Console.WriteLine("setting a = b1");
>    Foo<int>.SharedValue = "a2";

The WriteLine should be "setting a = a2"   :)

I'm looking at possibly making use of this project in some upcoming development so I have the follow tag set - sorry to bust in on the issue discussion...

**Comment by PeteGoo** [comment]
bust away! Welcome. The intention was to take advantage of the inherent per-type combination caching that comes as part of putting a static field on a generic.

Whoops, yes, updated the gist

**Comment by scottt732** [comment]
Yeah.. You're right.  I noticed the warning yesterday when I switched from the package reference to the library reference to debug through some stuff.  I was having a hell of a time debugging a bug where I was getting 1-3 of the events on an infinite Qbservable stream and naturally assumed it was your fault :-).  I was using a ReplaySubject with an AsQbservable and I needed to make them both static.  I was getting new instances of the context and didn't realize it.  That and I misinterpreted the Resharper warning backwardslywise.

So if the type collection really is specific to T's, nevermind that part.  I'm still trying to wrap my head all the way around Rx, but I think this IDisposable disposal thing might still be relevant.  

This project is a game changer...  It was pretty trivial to give a simple log4net appender splunk-like super powers.
